internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,570 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type PreAuthorizer interface {
+	PreAuthorize(ctx context.Context, manifestManager user.Info, manifestReader io.Reader) ([]ScopedPolicyRules, error)
+}
+
+type ScopedPolicyRules struct {
+	Namespace    string
+	MissingRules []rbacv1.PolicyRule
+}
+
+var objectVerbs = []string{"get", "patch", "update", "delete"}
+
+type rbacPreAuthorizer struct {
+	authorizer   authorizer.Authorizer
+	ruleResolver validation.AuthorizationRuleResolver
+	restMapper   meta.RESTMapper
+}
+
+func NewRBACPreAuthorizer(cl client.Client) PreAuthorizer {
+	return &rbacPreAuthorizer{
+		authorizer:   newRBACAuthorizer(cl),
+		ruleResolver: newRBACRulesResolver(cl),
+		restMapper:   cl.RESTMapper(),
+	}
+}
+
+// PreAuthorize validates whether the current user/request satisfies the necessary permissions
+// as defined by the RBAC policy. It examines the user’s roles, resource identifiers, and
+// the intended action to determine if the operation is allowed.
+//
+// Return Value:
+//   - nil: indicates that the authorization check passed and the operation is permitted.
+//   - non-nil error: indicates that an error occurred during the permission evaluation process
+//     (for example, a failure decoding the manifest or other internal issues). If the evaluation
+//     completes successfully but identifies missing rules, then a nil error is returned along with
+//     the list (or slice) of missing rules. Note that in some cases the error may encapsulate multiple
+//     evaluation failures
+func (a *rbacPreAuthorizer) PreAuthorize(ctx context.Context, manifestManager user.Info, manifestReader io.Reader) ([]ScopedPolicyRules, error) {
+	var allMissingPolicyRules = []ScopedPolicyRules{}
+	dm, err := a.decodeManifest(manifestReader)
+	if err != nil {
+		return nil, err
+	}
+	attributesRecords := dm.asAuthorizationAttributesRecordsForUser(manifestManager)
+
+	var preAuthEvaluationErrors []error
+	missingRules, err := a.authorizeAttributesRecords(ctx, attributesRecords)
+	if err != nil {
+		preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
+	}
+
+	ec := escalationChecker{
+		authorizer:        a.authorizer,
+		ruleResolver:      a.ruleResolver,
+		extraClusterRoles: dm.clusterRoles,
+		extraRoles:        dm.roles,
+	}
+
+	for _, obj := range dm.rbacObjects() {
+		if err := ec.checkEscalation(ctx, manifestManager, obj); err != nil {
+			missingEscalationRules, namespace := parseEscalationErrorForMissingRules(err)
+			// Check if we already have these escalation PolicyRules, if so don't append
+			for i, rule := range missingEscalationRules {
+				previousRule := missingRules[namespace][len(missingRules[namespace])-len(missingEscalationRules)+i]
+				if !arePolicyRulesEqual(previousRule, rule) {
+					missingRules[namespace] = append(missingRules[namespace], missingEscalationRules...)
+				}
+			}
+			preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
```

Comment:
Is `err` the original error here still (i.e. does it still include the missing rules text)? If so, I think we need `parseEscalationError` to return a new error where all the missing rules text is removed.

Reply:
ah yeah it's the original full error, so you just want to have it return the error minus the missing rules bit. can do

Reply:
Do you think it would be ... robust enough to split the error string at ": " a colon and a space?

Reply:
@bentito can close this thread